File: packages/core/src/delegate.ts
import type { RequestOptions, ViktorClient } from "./client.js";
import { ViktorError, ViktorRateLimitError } from "./errors.js";
import { DELEGATE_TOOL_SPEC } from "./generated/spec.js";

/** Name, description and JSON Schemas of the delegate tool. Single source: spec/delegate-tool.json. */
export const delegateToolSpec = DELEGATE_TOOL_SPEC;

export interface DelegateInput {
  task: string;
  thread_id?: string;
  response_schema?: Record<string, unknown>;
  speed?: "faster" | "smarter";
  timeout_seconds?: number;
}

export interface DelegateArtifact {
  display_name: string;
  content_type: string | null;
  download_url: string;
  expires_at: string | null;
}

export interface DelegateResult {
  status: "completed" | "requires_action" | "failed" | "cancelled" | "timed_out";
  markdown: string | null;
  json: unknown;
  artifacts: DelegateArtifact[];
  error: { code: string; message: string } | null;
  thread_id: string;
  run_id: string;
}

export interface DelegateOptions {
  signal?: AbortSignal;
  pollIntervalMs?: number;
  idempotencyKey?: string;
  /** Called on every poll with the current run status. */
  onStatus?: (status: string) => void;
  /** Resolve artifact download URLs (one extra request per artifact). Default: true. */
  resolveArtifacts?: boolean;
}

const ACTIVE = new Set(["queued", "in_progress", "cancellation_requested"]);
const REST = "/api/public/v1";

interface RunStatus {
  id: string;
  thread_id: string;
  status: string;
  error: { code: string; message: string } | null;
}

function sleep(ms: number, signal?: AbortSignal): Promise<void> {
  return new Promise((resolve, reject) => {
    if (signal?.aborted) return reject(signal.reason);
    const t = setTimeout(resolve, ms);
    signal?.addEventListener("abort", () => (clearTimeout(t), reject(signal.reason)), { once: true });
  });
}

/**
 * Delegate a task to Viktor over the native REST API: create a thread (or post to an existing
 * one), poll the run until it leaves the active states, then fetch the result and artifacts.
 * Unlike the chat wire this has no 600 s cap and reports `requires_action` explicitly.
 */
export async function delegateToViktor(client: ViktorClient, input: DelegateInput, options: DelegateOptions = {}): Promise<DelegateResult> {
  const { signal, pollIntervalMs = 2500, onStatus, resolveArtifacts = true } = options;
  const req: RequestOptions = { signal };
  const timeoutMs = (input.timeout_seconds ?? 600) * 1000;

  const body: Record<string, unknown> = {
    message: input.task,
    response_format: input.response_schema
      ? { type: "json_schema", json_schema: { name: "result", schema: input.response_schema } }
      : { type: "text" },
  };
  if (input.speed) body.speed = input.speed;

  const createOpts: RequestOptions = { ...req, idempotencyKey: options.idempotencyKey };
  let threadId: string;
  let runId: string;
  if (input.thread_id) {
    const created = await client.request<{ run: { id: string } }>("POST", `${REST}/threads/${encodeURIComponent(input.thread_id)}/messages`, body, createOpts);
    threadId = input.thread_id;
    runId = created.run.id;
  } else {
    const created = await client.request<{ thread: { id: string }; run: { id: string } }>("POST", `${REST}/threads`, body, createOpts);
    threadId = created.thread.id;
    runId = created.run.id;
  }

  const started = Date.now();
  let run: RunStatus;
  for (;;) {
    try {
      run = await client.request<RunStatus>("GET", `${REST}/runs/${encodeURIComponent(runId)}`, undefined, req);
    } catch (err) {
      if (err instanceof ViktorRateLimitError) {
        await sleep(Math.min((err.retryAfterSeconds ?? 5) * 1000, 30_000), signal);
        continue;
      }
      throw err;
    }
    onStatus?.(run.status);
    if (!ACTIVE.has(run.status)) break;
    if (Date.now() - started >= timeoutMs) {
      return { status: "timed_out", markdown: null, json: null, artifacts: [], error: null, thread_id: threadId, run_id: runId };
    }
    await sleep(pollIntervalMs, signal);
  }

  const base = { thread_id: threadId, run_id: runId };
  if (run.status === "cancelled") {
    return { status: "cancelled", markdown: null, json: null, artifacts: [], error: run.error, ...base };
  }

  let result: { status: string; markdown?: string | null; json?: unknown; artifacts?: Array<{ id: string; display_name?: string; content_type?: string | null }> };
  try {
    result = await client.request("GET", `${REST}/runs/${encodeURIComponent(runId)}/result`, undefined, req);
  } catch (err) {
    if (err instanceof ViktorError && err.status === 409) {
      const error = run.error ?? { code: err.detailCode ?? "result_not_available", message: err.message };
      return { status: run.status === "timed_out" ? "timed_out" : "failed", markdown: null, json: null, artifacts: [], error, ...base };
    }
    throw err;
  }

  const artifacts: DelegateArtifact[] = [];
  for (const a of result.artifacts ?? []) {
    let url = `${client.baseURL}${REST}/files/${encodeURIComponent(a.id)}/download-url`;
    let expires: string | null = null;
    if (resolveArtifacts) {
      try {
        const dl = await client.request<{ url: string; expires_at?: string }>("GET", `${REST}/files/${encodeURIComponent(a.id)}/download-url`, undefined, req);
        url = dl.url.startsWith("http") ? dl.url : `${client.baseURL}${dl.url}`;
        expires = dl.expires_at ?? null;
      } catch {
        /* keep the exchange URL; the caller can resolve it with the API key */
      }
    }
    artifacts.push({ display_name: a.display_name ?? a.id, content_type: a.content_type ?? null, download_url: url, expires_at: expires });
  }

  const status = run.status === "completed" || run.status === "requires_action" ? run.status : run.status === "timed_out" ? "timed_out" : "failed";
  return { status, markdown: result.markdown ?? null, json: result.json ?? null, artifacts, error: run.error, ...base } as DelegateResult;
}

/** Render a DelegateResult as the text a model should read as the tool result. */
export function formatDelegateResult(result: DelegateResult): string {
  const lines: string[] = [];
  if (result.status === "requires_action") lines.push("Viktor needs input before it can continue. Answer by calling this tool again with the same thread_id.");
  if (result.status === "timed_out") lines.push("Viktor is still working. Call this tool again later with the same thread_id to ask for the result.");
  if (result.status === "failed" || result.status === "cancelled") lines.push(`Viktor run ${result.status}${result.error ? `: ${result.error.message}` : "."}`);
  if (result.markdown) lines.push(result.markdown);
  if (result.json !== null && result.json !== undefined) lines.push("```json\n" + JSON.stringify(result.json, null, 2) + "\n```");
  for (const a of result.artifacts) lines.push(`File: ${a.display_name} (${a.content_type ?? "unknown type"}) ${a.download_url}`);
  lines.push(`(status: ${result.status}, thread_id: ${result.thread_id}, run_id: ${result.run_id})`);
  return lines.join("\n\n");
}

File: python/core/src/viktor_integrations_core/delegate.py
"""Delegate a task to Viktor over the native REST API (no 600 s cap, explicit requires_action, files)."""

from __future__ import annotations

import asyncio
import json
import time
from collections.abc import Callable
from dataclasses import asdict, dataclass, field
from typing import Any
from urllib.parse import quote

from ._spec import DELEGATE_TOOL_SPEC
from .client import AsyncViktorClient, ViktorClient
from .errors import ViktorError, ViktorRateLimitError

delegate_tool_spec: dict[str, Any] = DELEGATE_TOOL_SPEC
_ACTIVE = {"queued", "in_progress", "cancellation_requested"}
_REST = "/api/public/v1"


@dataclass
class DelegateResult:
    status: str
    thread_id: str
    run_id: str
    markdown: str | None = None
    json: Any = None
    artifacts: list[dict[str, Any]] = field(default_factory=list)
    error: dict[str, Any] | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    def to_text(self) -> str:
        """The text a model should read as the tool result."""
        lines: list[str] = []
        if self.status == "requires_action":
            lines.append(
                "Viktor needs input before it can continue. Answer by calling this tool again with the same thread_id."
            )
        if self.status == "timed_out":
            lines.append(
                "Viktor is still working. Call this tool again later with the same thread_id to ask for the result."
            )
        if self.status in ("failed", "cancelled"):
            lines.append(f"Viktor run {self.status}" + (f": {self.error['message']}" if self.error else "."))
        if self.markdown:
            lines.append(self.markdown)
        if self.json is not None:
            lines.append("```json\n" + json.dumps(self.json, indent=2) + "\n```")
        for a in self.artifacts:
            lines.append(f"File: {a['display_name']} ({a.get('content_type') or 'unknown type'}) {a['download_url']}")
        lines.append(f"(status: {self.status}, thread_id: {self.thread_id}, run_id: {self.run_id})")
        return "\n\n".join(lines)


def _create_request(task: str, thread_id: str | None, response_schema: dict[str, Any] | None, speed: str | None):
    body: dict[str, Any] = {
        "message": task,
        "response_format": (
            {"type": "json_schema", "json_schema": {"name": "result", "schema": response_schema}}
            if response_schema
            else {"type": "text"}
        ),
    }
    if speed:
        body["speed"] = speed
    path = f"{_REST}/threads/{quote(thread_id, safe='')}/messages" if thread_id else f"{_REST}/threads"
    return path, body


def _final_status(run_status: str) -> str:
    if run_status in ("completed", "requires_action", "cancelled", "timed_out"):
        return run_status
    return "failed"


def delegate_to_viktor(
    client: ViktorClient,
    task: str,
    *,
    thread_id: str | None = None,
    response_schema: dict[str, Any] | None = None,
    speed: str | None = None,
    timeout_seconds: float = 600,
    poll_interval: float = 2.5,
    idempotency_key: str | None = None,
    on_status: Callable[[str], None] | None = None,
    resolve_artifacts: bool = True,
) -> DelegateResult:
    path, body = _create_request(task, thread_id, response_schema, speed)
    created = client.request("POST", path, body, idempotency_key=idempotency_key)
    tid = thread_id or created["thread"]["id"]
    rid = created["run"]["id"]
    started = time.monotonic()
    while True:
        try:
            run = client.request("GET", f"{_REST}/runs/{quote(rid, safe='')}")
        except ViktorRateLimitError as exc:
            time.sleep(min(exc.retry_after_seconds or 5, 30))
            continue
        if on_status:
            on_status(run["status"])
        if run["status"] not in _ACTIVE:
            break
        if time.monotonic() - started >= timeout_seconds:
            return DelegateResult("timed_out", tid, rid)
        time.sleep(poll_interval)
    if run["status"] == "cancelled":
        return DelegateResult("cancelled", tid, rid, error=run.get("error"))
    try:
        result = client.request("GET", f"{_REST}/runs/{quote(rid, safe='')}/result")
    except ViktorError as exc:
        if exc.status == 409:
            error = run.get("error") or {"code": exc.detail_code or "result_not_available", "message": exc.message}
            return DelegateResult(
                _final_status(run["status"]) if run["status"] == "timed_out" else "failed", tid, rid, error=error
            )
        raise
    artifacts = []
    for a in result.get("artifacts") or []:
        url, expires = f"{client.base_url}{_REST}/files/{quote(a['id'], safe='')}/download-url", None
        if resolve_artifacts:
            try:
                dl = client.request("GET", f"{_REST}/files/{quote(a['id'], safe='')}/download-url")
                url = dl["url"] if dl["url"].startswith("http") else f"{client.base_url}{dl['url']}"
                expires = dl.get("expires_at")
            except ViktorError:
                pass
        artifacts.append(
            {
                "display_name": a.get("display_name") or a["id"],
                "content_type": a.get("content_type"),
                "download_url": url,
                "expires_at": expires,
            }
        )
    return DelegateResult(
        _final_status(run["status"]),
        tid,
        rid,
        markdown=result.get("markdown"),
        json=result.get("json"),
        artifacts=artifacts,
        error=run.get("error"),
    )


async def adelegate_to_viktor(
    client: AsyncViktorClient,
    task: str,
    *,
    thread_id: str | None = None,
    response_schema: dict[str, Any] | None = None,
    speed: str | None = None,
    timeout_seconds: float = 600,
    poll_interval: float = 2.5,
    idempotency_key: str | None = None,
    on_status: Callable[[str], None] | None = None,
    resolve_artifacts: bool = True,
) -> DelegateResult:
    path, body = _create_request(task, thread_id, response_schema, speed)
    created = await client.request("POST", path, body, idempotency_key=idempotency_key)
    tid = thread_id or created["thread"]["id"]
    rid = created["run"]["id"]
    started = time.monotonic()
    while True:
        try:
            run = await client.request("GET", f"{_REST}/runs/{quote(rid, safe='')}")
        except ViktorRateLimitError as exc:
            await asyncio.sleep(min(exc.retry_after_seconds or 5, 30))
            continue
        if on_status:
            on_status(run["status"])
        if run["status"] not in _ACTIVE:
            break
        if time.monotonic() - started >= timeout_seconds:
            return DelegateResult("timed_out", tid, rid)
        await asyncio.sleep(poll_interval)
    if run["status"] == "cancelled":
        return DelegateResult("cancelled", tid, rid, error=run.get("error"))
    try:
        result = await client.request("GET", f"{_REST}/runs/{quote(rid, safe='')}/result")
    except ViktorError as exc:
        if exc.status == 409:
            error = run.get("error") or {"code": exc.detail_code or "result_not_available", "message": exc.message}
            return DelegateResult("timed_out" if run["status"] == "timed_out" else "failed", tid, rid, error=error)
        raise
    artifacts = []
    for a in result.get("artifacts") or []:
        url, expires = f"{client.base_url}{_REST}/files/{quote(a['id'], safe='')}/download-url", None
        if resolve_artifacts:
            try:
                dl = await client.request("GET", f"{_REST}/files/{quote(a['id'], safe='')}/download-url")
                url = dl["url"] if dl["url"].startswith("http") else f"{client.base_url}{dl['url']}"
                expires = dl.get("expires_at")
            except ViktorError:
                pass
        artifacts.append(
            {
                "display_name": a.get("display_name") or a["id"],
                "content_type": a.get("content_type"),
                "download_url": url,
                "expires_at": expires,
            }
        )
    return DelegateResult(
        _final_status(run["status"]),
        tid,
        rid,
        markdown=result.get("markdown"),
        json=result.get("json"),
        artifacts=artifacts,
        error=run.get("error"),
    )
